Mathlib/Algebra/Lie/Weights.lean

```diff
@@ -249,6 +251,135 @@ instance [LieAlgebra.IsNilpotent R L] [IsNoetherian R M] :
     IsNilpotent R L (weightSpace M (0 : L → R)) :=
   isNilpotent_iff_forall'.mpr <| isNilpotent_toEndomorphism_weightSpace_zero M
 
+variable (R L)
+
+@[simp]
+lemma weightSpace_zero_normalizer_eq_self [LieAlgebra.IsNilpotent R L] :
+    (weightSpace M (0 : L → R)).normalizer = weightSpace M 0 := by
+  refine' le_antisymm _ (LieSubmodule.le_normalizer _)
+  intro m hm
+  rw [LieSubmodule.mem_normalizer] at hm
+  simp only [mem_weightSpace, Pi.zero_apply, zero_smul, sub_zero] at hm ⊢
+  intro y
+  obtain ⟨k, hk⟩ := hm y y
+  use k + 1
+  simpa [pow_succ', LinearMap.mul_eq_comp]
+
+lemma iSup_ucs_le_weightSpace_zero [LieAlgebra.IsNilpotent R L] :
+    ⨆ k, (⊥ : LieSubmodule R L M).ucs k ≤ weightSpace M (0 : L → R) := by
+  simpa using LieSubmodule.ucs_le_of_normalizer_eq_self (weightSpace_zero_normalizer_eq_self R L M)
+
+/-- See also `LieModule.iInf_lowerCentralSeries_eq_posFittingComp`. -/
+lemma iSup_ucs_eq_weightSpace_zero [LieAlgebra.IsNilpotent R L] [IsNoetherian R M] :
+    ⨆ k, (⊥ : LieSubmodule R L M).ucs k = weightSpace M (0 : L → R) := by
+  obtain ⟨k, hk⟩ := (LieSubmodule.isNilpotent_iff_exists_self_le_ucs
+    <| weightSpace M (0 : L → R)).mp inferInstance
+  refine le_antisymm (iSup_ucs_le_weightSpace_zero R L M) (le_trans hk ?_)
+  exact le_iSup (fun k ↦ (⊥ : LieSubmodule R L M).ucs k) k
+
+variable {L}
+
+/-- If `M` is a representation of a nilpotent Lie algebra `L`, and `x : L`, then
+`posFittingCompOf R M x` is the positive Fitting component of the action of `x` on `M`.
+
+It is a Lie submodule because `L` is nilpotent. -/
+def posFittingCompOf [LieAlgebra.IsNilpotent R L] (x : L) : LieSubmodule R L M :=
+  { ⨅ k, LinearMap.range (toEndomorphism R L M x ^ k) with
```

Comment:
A couple of months ago I was terrified by `with` and was removing them from mathlib and making stuff faster. Now Chris Hughes has started adding them in other places and making stuff faster. If you `#print` this definition then you can see a (possibly formerly) dreaded `let src := ...`. Do you understand what's going on well enough to know whether this definition is a better choice than `toSubmodule := ⨅ k, LinearMap.range (toEndomorphism R L M x ^ k), lie_mem := ...`? I don't (and I wish I did).

Reply:
Thanks, I do not understand this any better but I certainly prefer the term generated by `toSubmodule :=` rather than ` with` so I have applied this suggestion.